### PR TITLE
[dynamo 3.11] changes to with contexts

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94243
* #94102
* __->__ #94101
* #94100
* #94099
* #94098

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/symbolic_convert.py [hunk 1]
```diff
@@ -1558,6 +1561,13 @@ def CALL(self, inst):
             kwargs = {}
         self.call_function(fn, args, kwargs)
         self.kw_names = None
+        # 3.11 removed POP_BLOCK, so we manually pop the block stack here
+        if (
+            isinstance(fn, WithExitFunctionVariable)
+            and len(self.block_stack) > 0
+            and id(fn) == self.block_stack[-1].id
```
Comment (albanD) [inline torch/_dynamo/symbolic_convert.py:1568]: Could there be more than one associated with this function?

File: torch/_dynamo/symbolic_convert.py [hunk 2]
```diff
@@ -897,7 +900,7 @@ def SETUP_WITH(self, inst):
         self.push(ctx.enter(self))
 
     def SETUP_FINALLY(self, inst):
-        self.block_stack.append(BlockStackEntry(inst.target))
+        self.block_stack.append(BlockStackEntry(0, inst.target))
```
Comment (albanD) [inline torch/_dynamo/symbolic_convert.py:903]: For all these functions that don't exist for 3.11, should we comment them the same way the ones only existing for python<=3.7 are?

**Review by albanD — COMMENTED** [review]
I was expecting to see some fix of the exception table alongside the new cleanup code?

**Comment by williamwen42** [comment]
@albanD I haven't had time to take a look at exception tables yet - I was planning on fixing it at the top of the PR stack.

**Comment by albanD** [comment]
Delaying the exception table change sounds ok to me!

**Review by albanD — APPROVED** [review]
Let's get this in!

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**